README.md
⛔️ DEPRECATED: This module has been superseded by [@ipld/dag-cbor](https://github.com/ipld/js-dag-cbor) and [multiformats](https://github.com/multiformats/js-multiformats) <!-- omit in toc -->
======

# js-ipld-dag-cbor <!-- omit in toc -->

[![](https://img.shields.io/badge/made%20by-Protocol%20Labs-blue.svg?style=flat-square)](http://ipn.io)
[![](https://img.shields.io/badge/project-IPLD-blue.svg?style=flat-square)](http://github.com/ipld/ipld)
[![](https://img.shields.io/badge/freenode-%23ipfs-blue.svg?style=flat-square)](http://webchat.freenode.net/?channels=%23ipfs)
[![Travis CI](https://flat.badgen.net/travis/ipld/js-ipld-dag-cbor)](https://travis-ci.com/ipld/js-ipld-dag-cbor)
[![Coverage](https://coveralls.io/repos/github/ipld/js-ipld-dag-cbor/badge.svg?branch=master)](https://coveralls.io/github/ipld/js-ipld-dag-cbor?branch=master)
[![](https://img.shields.io/badge/standard--readme-OK-green.svg?style=flat-square)](https://github.com/RichardLitt/standard-readme)
[![](https://david-dm.org/ipld/js-ipld-dag-cbor.svg?style=flat-square)](https://david-dm.org/ipld/js-ipld-dag-cbor)
[![](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/feross/standard)
[![Greenkeeper badge](https://badges.greenkeeper.io/ipld/js-ipld-dag-cbor.svg)](https://greenkeeper.io/)
![](https://img.shields.io/badge/npm-%3E%3D3.0.0-orange.svg?style=flat-square)
![](https://img.shields.io/badge/Node.js-%3E%3D6.0.0-orange.svg?style=flat-square)

> JavaScript implementation of the [IPLD spec](https://github.com/ipfs/specs/tree/master/ipld).

## Lead Maintainer <!-- omit in toc -->

[Volker Mische](https://github.com/vmx)

## Table of Contents <!-- omit in toc -->

- [Install](#install)
  - [npm](#npm)
  - [Use in Node.js](#use-in-nodejs)
  - [Use in a browser with browserify, webpack or any other bundler](#use-in-a-browser-with-browserify-webpack-or-any-other-bundler)
  - [Use in a browser Using a script tag](#use-in-a-browser-using-a-script-tag)
- [Usage](#usage)
- [API](#api)
  - [`dagCBOR.util.serialize(obj)`](#dagcborutilserializeobj)
  - [`dagCBOR.util.deserialize(serialized)`](#dagcborutildeserializeserialized)
  - [`dagCBOR.util.configureDecoder([options])`](#dagcborutilconfiguredecoderoptions)
  - [`dagCBOR.util.cid(obj[, options,])`](#dagcborutilcidobj-options)
- [Contribute](#contribute)
- [License](#license)

## Install

### npm

```sh
> npm install ipld-dag-cbor
```

### Use in Node.js

```JavaScript
const dagCBOR = require('ipld-dag-cbor')
```

### Use in a browser with browserify, webpack or any other bundler

The code published to npm that gets loaded on require is in fact a ES5 transpiled version with the right shims added. This means that you can require it and use with your favourite bundler without having to adjust asset management process.

```JavaScript
var dagCBOR = require('ipld-dag-cbor')
```

### Use in a browser Using a script tag

Loading this module through a script tag will make the `IpldDagCbor` obj available in the global namespace.

```html
<script src="https://unpkg.com/ipld-dag-cbor/dist/index.min.js"></script>
<!-- OR -->
<script src="https://unpkg.com/ipld-dag-cbor/dist/index.js"></script>
```

## Usage

```JavaScript
const dagCBOR = require('ipld-dag-cbor')

const file = {
  name: 'hello.txt',
  size: 11
}

const serialized = dagCBOR.util.serialize(file)
console.log(`Encoded as a ${serialized.length} byte Uint8Array`)

const node = dagCBOR.util.deserialize(serialized)
console.log('Decoded as:', node)
require('assert').deepEqual(node, file) // should match

// → Encoded as a 22 byte Uint8Array
// → Decoded as: { name: 'hello.txt', size: 11 }
```

## API

### `dagCBOR.util.serialize(obj)`

Encodes an object into IPLD CBOR form, replacing any CIDs found within the object to CBOR tags (with an id of `42`).

 - `obj` (any): any object able to be serialized as CBOR

Returns the serialized node.

### `dagCBOR.util.deserialize(serialized)`

 Decodes an IPLD CBOR encoded representation, restoring any CBOR tags (id `42`) to CIDs.

  - `serialized` (`Uint8Array` or `String`): a binary blob representing an IPLD CBOR encoded object.

Returns the deserialized object.

### `dagCBOR.util.configureDecoder([options])`

Configure the underlying CBOR decoder.

Possible values in the `options` argument are:

 - `size` (`Number`, optional): the current heap size used in CBOR parsing, this may grow automatically as larger blocks are encountered up to `maxSize`. Default: `65536` (64Kb).
 - `maxSize` (`Number`, optional): the maximum size the CBOR parsing heap is allowed to grow to before `dagCBOR.util.deserialize()` returns an error. Default: `67108864` (64Mb).
 - `tags` (`Object`, optional): an object whose keys are CBOR tag numbers and values are transform functions that accept a `value` and return a decoded representation of that `value`.

 The CBOR decoder uses a heap size that is a power of two. Setting `size` to a number other than a power of two will result in a heap using the next-largest power of two.

 Calling `dagCBOR.util.configureDecoder()` with no arguments will reset to the default decoder `size`, `maxSize` and `tags`.

### `dagCBOR.util.cid(obj[, options,])`

Create a [CID](https://github.com/multiformats/js-cid) for the given unserialized object.

 - `obj` (any): any object able to be serialized as CBOR
 - `options` (`Object`):
  * `hashAlg` (`String`): a [registered multicodec](https://github.com/multiformats/multicodec/blob/master/table.csv) hash algorithm.
  * `hashLen` (`String`): an optional hash length
  * `version` (`Number`): CID version number, defaults to `1`

Returns a Promise with the created CID.

## Contribute

Feel free to join in. All welcome. Open an [issue](https://github.com/ipld/js-ipld-dag-cbor/issues)!

Check out our [contributing document](https://github.com/ipld/ipld/blob/master/contributing.md) for more information on how we work, and about contributing in general. Please be aware that all interactions related to IPLD are subject to the IPFS [Code of Conduct](https://github.com/ipfs/community/blob/master/code-of-conduct.md).

Small note: If editing the README, please conform to the [standard-readme](https://github.com/RichardLitt/standard-readme) specification.

## License

[MIT](LICENSE) © 2015 David Dias


CHANGELOG.md
## [1.0.1](https://github.com/ipld/js-ipld-dag-cbor/compare/v1.0.0...v1.0.1) (2021-08-11)



# [1.0.0](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.18.0...v1.0.0) (2021-04-27)


### chore

* update deps ([#140](https://github.com/ipld/js-ipld-dag-cbor/issues/140)) ([5d13035](https://github.com/ipld/js-ipld-dag-cbor/commit/5d13035533a00fba12d4a0e9544347cce51e1350))


### BREAKING CHANGES

* `buffer@6.x.x` has dropped support for IE and Safari < 10



# [0.18.0](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.17.1...v0.18.0) (2021-03-11)



## [0.17.1](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.17.0...v0.17.1) (2021-03-03)



<a name="0.17.0"></a>
# [0.17.0](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.16.0...v0.17.0) (2020-08-04)


### Bug Fixes

* replace node buffers with uint8arrays ([#134](https://github.com/ipld/js-ipld-dag-cbor/issues/134)) ([41587e7](https://github.com/ipld/js-ipld-dag-cbor/commit/41587e7))


### BREAKING CHANGES

* - `util.serialize` now returns a Uint8Array



<a name="0.16.0"></a>
# [0.16.0](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.15.3...v0.16.0) (2020-07-10)


### Bug Fixes

* reject CBOR data with extraneous back-to-back encoded data ([8176c13](https://github.com/ipld/js-ipld-dag-cbor/commit/8176c13)), closes [/github.com/dignifiedquire/borc/issues/47#issuecomment-642432759](https://github.com//github.com/dignifiedquire/borc/issues/47/issues/issuecomment-642432759)



<a name="0.15.3"></a>
## [0.15.3](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.15.2...v0.15.3) (2020-06-10)


### Features

* serialize Uint8Array as binary (like Buffer) ([95f6bda](https://github.com/ipld/js-ipld-dag-cbor/commit/95f6bda))



<a name="0.15.2"></a>
## [0.15.2](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.15.1...v0.15.2) (2020-03-27)


### Bug Fixes

* remove node globals ([#124](https://github.com/ipld/js-ipld-dag-cbor/issues/124)) ([7acfeae](https://github.com/ipld/js-ipld-dag-cbor/commit/7acfeae))
* **package:** update cids to version 0.8.0 ([bc6ac54](https://github.com/ipld/js-ipld-dag-cbor/commit/bc6ac54))



<a name="0.15.1"></a>
## [0.15.1](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.15.0...v0.15.1) (2020-01-13)


### Bug Fixes

* **package:** update multicodec to version 1.0.0 ([d1d78a0](https://github.com/ipld/js-ipld-dag-cbor/commit/d1d78a0))
* **package:** update multihashing-async to version 0.8.0 ([b8d4a55](https://github.com/ipld/js-ipld-dag-cbor/commit/b8d4a55))



<a name="0.15.0"></a>
# [0.15.0](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.14.0...v0.15.0) (2019-05-10)


### Bug Fixes

* **package:** update cids to version 0.7.0 ([2dc3378](https://github.com/ipld/js-ipld-dag-cbor/commit/2dc3378))


### BREAKING CHANGES

* **package:** Returned v1 CIDs now default to base32 encoding

Previous versions returned a base58 encoded string when `toString()`/
`toBaseEncodedString()` was called on a CIDv1. It now returns a base32
encoded string.



<a name="0.14.0"></a>
# [0.14.0](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.13.1...v0.14.0) (2019-05-08)


### Bug Fixes

* make cbor Decoder configurable ([#90](https://github.com/ipld/js-ipld-dag-cbor/issues/90)) ([dfb9137](https://github.com/ipld/js-ipld-dag-cbor/commit/dfb9137))
* remove console.log statement ([a413cb2](https://github.com/ipld/js-ipld-dag-cbor/commit/a413cb2))
* typo in travis badge ([94122f0](https://github.com/ipld/js-ipld-dag-cbor/commit/94122f0))
* **package:** update cids to version 0.6.0 ([1d507f7](https://github.com/ipld/js-ipld-dag-cbor/commit/1d507f7))
* **package:** update multihashing-async to version 0.6.0 ([b328072](https://github.com/ipld/js-ipld-dag-cbor/commit/b328072))


### Features

* allow decoder heap to grow dynamically ([1f7b7f1](https://github.com/ipld/js-ipld-dag-cbor/commit/1f7b7f1)), closes [#73](https://github.com/ipld/js-ipld-dag-cbor/issues/73)
* new IPLD Format API ([cfc8519](https://github.com/ipld/js-ipld-dag-cbor/commit/cfc8519))


### BREAKING CHANGES

* The API is now async/await based

There are numerous changes, the most significant one is that the API
is no longer callback based, but it using async/await.

For the full new API please see the [IPLD Formats spec].

[IPLD Formats spec]: https://github.com/ipld/interface-ipld-format



<a name="0.13.1"></a>
## [0.13.1](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.13.0...v0.13.1) (2019-01-08)


### Bug Fixes

* reduce bundle size ([#87](https://github.com/ipld/js-ipld-dag-cbor/issues/87)) ([ab637f3](https://github.com/ipld/js-ipld-dag-cbor/commit/ab637f3))


### Features

* support the hashLen option of multihashing ([#83](https://github.com/ipld/js-ipld-dag-cbor/issues/83)) ([9ffb5e2](https://github.com/ipld/js-ipld-dag-cbor/commit/9ffb5e2))



<a name="0.13.0"></a>
# [0.13.0](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.12.1...v0.13.0) (2018-10-01)


### Bug Fixes

* resolve falsy values ([0a49705](https://github.com/ipld/js-ipld-dag-cbor/commit/0a49705))


### Features

* serialize and de-serialize CID instances ([8585d65](https://github.com/ipld/js-ipld-dag-cbor/commit/8585d65))


### BREAKING CHANGES

* return values from de-serializer are now CID instances.
Serializer still supports old link objects.



<a name="0.12.1"></a>
## [0.12.1](https://github.com/ipld/js-ipld-dag-cbor/compare/v0.12.0...v0.12.1) (2018-06-29)


### Bug Fixes

* pass serialized blob to util.cid ([#67](https://github.com/ipld/js-ipld-dag-cbor/issues/67)) ([1ec7744](https://github.com/ipld/js-ipld-dag-cbor/commit/1ec7744))


### Features

* add defaultHashAlg ([#65](https://github.com/ipld/js-ipld-dag-cbor/issues/65)) ([e095ef5](https://github.com/ipld/js-ipld-dag-cbor/commit/e095ef5))
* add util.cid options ([#66](https://github.com/ipld/js-ipld-dag-cbor/issues/66)) ([1aed60e](https://github.com/ipld/js-ipld-dag-cbor/commit/1aed60e))


### BREAKING CHANGES

* the first argument is now the serialized output NOT the dag node.

test/mod.spec.js
/* eslint-env mocha */
'use strict'

const { expect } = require('aegir/utils/chai')
const multicodec = require('multicodec')
const mod = require('../src')

describe('IPLD Format', () => {
  it('codec is dag-cbor', () => {
    expect(mod.codec).to.equal(multicodec.DAG_CBOR)
  })

  it('defaultHashAlg is sha2-256', () => {
    expect(mod.defaultHashAlg).to.equal(multicodec.SHA2_256)
  })
})


test/cbor.spec.js
/* eslint-env mocha */
'use strict'

const expect = require('chai').expect
const cbor = require('cbor')
const Multiaddr = require('multiaddr')

const ipld = require('../src')

describe('IPLD -> CBOR', () => {
  it('no /', () => {
    const src = {
      data: 'hello world',
      size: 11
    }

    const expected = {
      data: 'hello world',
      size: 11
    }

    expect(
      ipld.marshal(src)
    ).to.be.eql(
      cbor.encode(expected)
    )
  })

  it('/, is a string', () => {
    const src = {
      data: 'hello world',
      size: 11,
      l1: {'/': 'hello-world'}
    }

    const expected = {
      data: 'hello world',
      size: 11,
      l1: new cbor.Tagged(ipld.LINK_TAG, 'hello-world')
    }

    expect(
      ipld.marshal(src)
    ).to.be.eql(
      cbor.encode(expected)
    )
  })

  it('/, is a valid multiaddress', () => {
    const addr1 = new Multiaddr('/ip4/127.0.0.1/udp/1234')
    const addr2 = new Multiaddr('/ipfs/Qmafmh1Cw3H1bwdYpaaj5AbCW4LkYyUWaM7Nykpn5NZoYL')
    const src = {
      data: 'hello world',
      size: 11,
      l1: {'/': addr1.toString()},
      l2: {'/': addr2.toString()}
    }

    const expected = {
      data: 'hello world',
      size: 11,
      l1: new cbor.Tagged(ipld.LINK_TAG, addr1.buffer),
      l2: new cbor.Tagged(ipld.LINK_TAG, addr2.buffer)
    }

    expect(
      ipld.marshal(src)
    ).to.be.eql(
      cbor.encode(expected)
    )
  })

  it('/, with properties', () => {
    const src = {
      data: 'hello world',
      size: 11,
      secret: {
        '/': 'hello-world',
        i: 'should not be here'
      }
    }

    expect(
      () => ipld.marshal(src)
    ).to.throw(
      'Links must not have siblings'
    )
  })

  it('nested /', () => {
    const src = {
      data: 'hello world',
      size: 11,
      l1: {'/': 'hello-world'},
      secret: {
        l1: {'/': 'secret-link'}
      }
    }

    const expected = {
      data: 'hello world',
      size: 11,
      l1: new cbor.Tagged(ipld.LINK_TAG, 'hello-world'),
      secret: {
        l1: new cbor.Tagged(ipld.LINK_TAG, 'secret-link')
      }
    }

    expect(
      ipld.marshal(src)
    ).to.be.eql(
      cbor.encode(expected)
    )
  })

  it('does not modify the input', () => {
    let src = {
      l1: {'/': 'hello'}
    }

    ipld.marshal(src)

    expect(src).to.be.eql({
      l1: {'/': 'hello'}
    })
  })

  it('marshals objects with null values', () => {
    const src = {
      l1: {'/': 'hello'},
      foo: null
    }

    const expected = {
      l1: new cbor.Tagged(ipld.LINK_TAG, 'hello'),
      foo: null
    }

    expect(
      ipld.marshal(src)
    ).to.be.eql(
      cbor.encode(expected)
    )
  })
})

describe('CBOR -> IPLD', () => {
  it('no links', () => {
    const src = cbor.encode({
      data: 'hello world',
      size: 11
    })

    const target = {
      data: 'hello world',
      size: 11
    }

    expect(
      ipld.unmarshal(src)
    ).to.be.eql(
      target
    )
  })

  it('one link, without properties', () => {
    const src = cbor.encode({
      data: 'hello world',
      size: 11,
      nested: new cbor.Tagged(ipld.LINK_TAG, 'hello-world')
    })

    const target = {
      data: 'hello world',
      size: 11,
      nested: {
        '/': 'hello-world'
      }
    }

    expect(
      ipld.unmarshal(src)
    ).to.be.eql(
      target
    )
  })
})


test/test.spec.js
 /* eslint-env mocha */
'use strict'

var expect = require('chai').expect
var ipld = require('../src')

describe('IPLD Tests', (done) => {
  it('add the mlink as context a json blob', function (done) {
    var node = {
      data: 'aaah the data'
    }

    node['@context'] = ipld.context.merkleweb

    var expected = {
      data: 'aaah the data',
      '@context': {
        mlink: 'http://merkle-link'
      }
    }

    expect(node).to.deep.equal(expected)
    done()
  })

  it('expand', function (done) {
    var node = {
      data: 'aaah the data',
      mlink: 'GIMETHEHASSSSSH'
    }

    node['@context'] = ipld.context.merkleweb

    var expected = {
      data: 'aaah the data',
      'http://merkle-link': 'GIMETHEHASSSSSH'
    }

    expect(ipld.expand(node)).to.deep.equal(expected)
    done()
  })

  it('expand with buffer as val', function (done) {
    var node = {
      data: 'aaah the data',
      mlink: new Buffer('GIVEMETHEHASH')
    }

    node['@context'] = ipld.context.merkleweb

    var expected = {
      data: 'aaah the data',
      'http://merkle-link': new Buffer('GIVEMETHEHASH')
    }

    expect(ipld.expand(node)).to.deep.equal(expected)
    done()
  })

  it('expand nested', function (done) {
    var node = {
      data: 'aaah the data',
      foo: {
        mlink: 'GIMETHEHASSSSSH'
      }
    }

    node['@context'] = ipld.context.merkleweb

    var expected = {
      data: 'aaah the data',
      foo: {
        'http://merkle-link': 'GIMETHEHASSSSSH'
      }
    }

    expect(ipld.expand(node)).to.deep.equal(expected)
    done()
  })

  it('marshel and unmarshal', function (done) {
    var node = {
      data: 'aaah the data',
      mlink: 'GIMETHEHASSSSSH'
    }

    node['@context'] = ipld.context.merkleweb

    ipld.unmarshalAsync(ipld.marshal(node), function (err, result) {
      expect(err).to.equal(null)
      var expected = {
        data: 'aaah the data',
        'http://merkle-link': 'GIMETHEHASSSSSH'
      }

      expect(ipld.expand(result[0])).to.deep.equal(expected)
      done()
    })
  })

  it('marshel and unmarshal Sync', function (done) {
    var node = {
      data: 'aaah the data',
      mlink: 'GIMETHEHASSSSSH'
    }

    node['@context'] = ipld.context.merkleweb

    var result = ipld.unmarshal(ipld.marshal(node))

    var expected = {
      data: 'aaah the data',
      'http://merkle-link': 'GIMETHEHASSSSSH'
    }

    expect(ipld.expand(result)).to.deep.equal(expected)
    done()
  })
})


src/cbor.js
'use strict'

const cbor = require('cbor')
const cborUtils = require('cbor/lib/utils')
const Multiaddr = require('multiaddr')
const NoFilter = require('nofilter')
const defaults = require('lodash.defaults')
const includes = require('lodash.includes')
const cloneDeep = require('lodash.clonedeep')

exports = module.exports

exports.LINK_TAG = 258
const LINK_SYMBOL = exports.LINK_SYMBOL = '/'

exports.marshal = (original) => {
  const input = cloneDeep(original)

  function transform (obj) {
    if (obj == null) return obj
    const keys = Object.keys(obj)

    // Recursive transform
    keys.forEach((key) => {
      if (typeof obj[key] === 'object') {
        obj[key] = transform(obj[key])
      }
    })

    if (includes(keys, LINK_SYMBOL)) {
      let link = obj[LINK_SYMBOL]

      // Multiaddr encoding
      if (typeof link === 'string' && isMultiaddr(link)) {
        link = new Multiaddr(link).buffer
      }

      // Remove the @link
      delete obj[LINK_SYMBOL]

      // Non empty
      if (keys.length > 1) {
        throw new Error('Links must not have siblings')
      }

      return new cbor.Tagged(exports.LINK_TAG, link)
    }

    return obj
  }

  return cbor.encode(transform(input))
}

exports.unmarshal = (input, opts) => {
  opts = defaults(opts || {}, {
    encoding: cborUtils.guessEncoding(input)
  })

  const dec = new cbor.Decoder(opts)
  const bs = new NoFilter()

  dec.pipe(bs)
  dec.end(input, opts.encoding)

  const res = bs.read()

  function transform (obj) {
    Object.keys(obj).forEach((key) => {
      const val = obj[key]
      // This is safe as we reference the same cbor instance
      // as we used to decode with
      if (val instanceof cbor.Tagged) {
        if (typeof val.value === 'string') {
          obj[key] = {
            [LINK_SYMBOL]: val.value
          }
        } else if (Buffer.isBuffer(val.value)) {
          obj[key] = {
            [LINK_SYMBOL]: (new Multiaddr(val.value)).toString()
          }
        } else {
          obj[key] = defaults({
            [LINK_SYMBOL]: val.value[0]
          }, transform(val.value[1]))
        }
      } else if (typeof val === 'object') {
        obj[key] = transform(val)
      }
    })

    return obj
  }

  return transform(res)
}

function isMultiaddr (str) {
  try {
    const addr = new Multiaddr(str)
    return addr.toString() === str
  } catch (err) {
    return false
  }
}


src/resolver.js
'use strict'

const CID = require('cids')
const util = require('./util')

/**
 * Resolves a path within a CBOR block.
 *
 * Returns the value or a link and the partial mising path. This way the
 * IPLD Resolver can fetch the link and continue to resolve.
 *
 * @param {Uint8Array} binaryBlob - Binary representation of a CBOR block
 * @param {string} [path='/'] - Path that should be resolved
 */
exports.resolve = (binaryBlob, path = '/') => {
  let node = util.deserialize(binaryBlob)

  const parts = path.split('/').filter(Boolean)
  while (parts.length) {
    const key = parts.shift()
    if (!key || !(key in node)) {
      throw new Error(`Object has no property '${key}'`)
    }

    node = node[key]
    if (CID.isCID(node)) {
      return {
        value: node,
        remainderPath: parts.join('/')
      }
    }
  }

  return {
    value: node,
    remainderPath: ''
  }
}

/**
 * @param {any} node
 * @param {string} [path]
 * @returns {Generator<string, void, undefined>}
 */
const traverse = function * (node, path) {
  // Traverse only objects and arrays
  if (node instanceof Uint8Array || CID.isCID(node) || typeof node === 'string' || node === null) {
    return
  }

  for (const item of Object.keys(node)) {
    const nextpath = path === undefined ? item : path + '/' + item
    yield nextpath
    yield * traverse(node[item], nextpath)
  }

  // to stop eslint and tsc fighting
  return undefined
}

/**
 * Return all available paths of a block.
 *
 * @generator
 * @param {Uint8Array} binaryBlob - Binary representation of a CBOR block
 * @yields {string} - A single path
 */
exports.tree = function * (binaryBlob) {
  const node = util.deserialize(binaryBlob)

  yield * traverse(node)
}


src/util.js
'use strict'

// @ts-ignore TODO: switch to cborg
const cbor = require('borc')
const multicodec = require('multicodec')
const multihashing = require('multihashing-async')
const { multihash } = multihashing
const CID = require('cids')
// @ts-ignore
const isCircular = require('is-circular')
const uint8ArrayConcat = require('uint8arrays/concat')
const uint8ArrayFromString = require('uint8arrays/from-string')

/**
 * @typedef {import('cids').CIDVersion} CIDVersion
 * @typedef {import('multihashing-async').multihash.HashCode} HashCode
 */

// https://github.com/ipfs/go-ipfs/issues/3570#issuecomment-273931692
const CID_CBOR_TAG = 42

/**
 * @param {CID | string} cid
 */
function tagCID (cid) {
  let buf

  if (typeof cid === 'string') {
    buf = new CID(cid).bytes
  } else if (CID.isCID(cid)) {
    buf = cid.bytes
  } else {
    throw new Error('Could not tag CID - was not string or CID')
  }

  return new cbor.Tagged(CID_CBOR_TAG, uint8ArrayConcat([
    uint8ArrayFromString('00', 'base16'), // thanks jdag
    buf
  ], 1 + buf.length))
}

/**
 * @param {any} dagNode
 */
function replaceCIDbyTAG (dagNode) {
  let circular
  try {
    circular = isCircular(dagNode)
  } catch (e) {
    circular = false
  }
  if (circular) {
    throw new Error('The object passed has circular references')
  }

  /**
   * @param {any} obj
   * @returns {any}
   */
  function transform (obj) {
    if (!obj || obj instanceof Uint8Array || typeof obj === 'string') {
      return obj
    }

    if (Array.isArray(obj)) {
      return obj.map(transform)
    }

    if (CID.isCID(obj)) {
      return tagCID(obj)
    }

    const keys = Object.keys(obj)

    if (keys.length > 0) {
      // Recursive transform
      /** @type {Record<string, any>} */
      const out = {}
      keys.forEach((key) => {
        if (typeof obj[key] === 'object') {
          out[key] = transform(obj[key])
        } else {
          out[key] = obj[key]
        }
      })
      return out
    } else {
      return obj
    }
  }

  return transform(dagNode)
}

const codec = multicodec.DAG_CBOR
const defaultHashAlg = multihash.names['sha2-256']

const defaultTags = {
  /**
   * @param {Uint8Array} val
   */
  [CID_CBOR_TAG]: (val) => {
    // remove that 0
    val = val.slice(1)
    return new CID(val)
  }
}
const defaultSize = 64 * 1024 // current decoder heap size, 64 Kb
let currentSize = defaultSize
const defaultMaxSize = 64 * 1024 * 1024 // max heap size when auto-growing, 64 Mb
let maxSize = defaultMaxSize
/** @type {cbor.Decoder} */
let decoder

/**
 * Configure the underlying CBOR decoder.
 *
 * @param {Object} [options] - The options the decoder takes. The decoder will reset to the defaul values if no options are given.
 * @param {number} [options.size=65536] - The current heap size used in CBOR parsing, this may grow automatically as larger blocks are encountered up to `maxSize`
 * @param {number} [options.maxSize=67108864] - The maximum size the CBOR parsing heap is allowed to grow to before `dagCBOR.util.deserialize()` returns an error
 * @param {Object} [options.tags] - An object whose keys are CBOR tag numbers and values are transform functions that accept a `value` and return a decoded representation of that `value`
 */
function configureDecoder (options) {
  let tags = defaultTags

  if (options) {
    if (typeof options.size === 'number') {
      currentSize = options.size
    }
    if (typeof options.maxSize === 'number') {
      maxSize = options.maxSize
    }
    if (options.tags) {
      tags = Object.assign({}, defaultTags, options && options.tags)
    }
  } else {
    // no options, reset to defaults
    currentSize = defaultSize
    maxSize = defaultMaxSize
  }

  const decoderOptions = {
    tags,
    size: currentSize
  }

  decoder = new cbor.Decoder(decoderOptions)
  // borc edits opts.size in-place so we can capture _actual_ size
  currentSize = decoderOptions.size
}

configureDecoder() // Setup default cbor.Decoder

/**
 * Serialize internal representation into a binary CBOR block.
 *
 * @param {Object} node - Internal representation of a CBOR block
 * @returns {Uint8Array} - The encoded binary representation
 */
function serialize (node) {
  const nodeTagged = replaceCIDbyTAG(node)
  const serialized = cbor.encode(nodeTagged)

  return serialized
}

/**
 * Deserialize CBOR block into the internal representation.
 *
 * @param {Uint8Array} data - Binary representation of a CBOR block
 * @returns {{ [key: string]: any }} - An object that conforms to the IPLD Data Model
 */
function deserialize (data) {
  if (data.length > currentSize && data.length <= maxSize) {
    configureDecoder({ size: data.length })
  }

  if (data.length > currentSize) {
    throw new Error('Data is too large to deserialize with current decoder')
  }

  // borc will decode back-to-back objects into an implicit top-level array, we
  // strictly want to only see a single explicit top-level object
  const all = decoder.decodeAll(data)
  if (all.length !== 1) {
    throw new Error('Extraneous CBOR data found beyond initial top-level object')
  }

  return all[0]
}

/**
 * Calculate the CID of the binary blob.
 *
 * @param {Uint8Array} binaryBlob - Encoded IPLD Node
 * @param {Object} [userOptions] - Options to create the CID
 * @param {CIDVersion} [userOptions.cidVersion=1] - CID version number
 * @param {HashCode} [userOptions.hashAlg=multihash.names['sha2-256']] - Defaults to the defaultHashAlg of the format
 */
async function cid (binaryBlob, userOptions = {}) {
  const options = {
    cidVersion: userOptions.cidVersion == null ? 1 : userOptions.cidVersion,
    hashAlg: userOptions.hashAlg == null ? module.exports.defaultHashAlg : userOptions.hashAlg
  }

  const hashName = multihash.codes[options.hashAlg]
  const hash = await multihashing(binaryBlob, hashName)
  const codecName = multicodec.getNameFromCode(module.exports.codec)
  const cid = new CID(options.cidVersion, codecName, hash)

  return cid
}

module.exports = {
  codec,
  defaultHashAlg,
  configureDecoder,
  serialize,
  deserialize,
  cid
}
